examples/advanced/Tutorial3/reconstruction/FairTestDetectorRecoTask.cxx

```diff
@@ -97,7 +97,9 @@ void FairTestDetectorRecoTask::Exec(Option_t* /*opt*/)
         TVector3 dpos(1 / TMath::Sqrt(12), 1 / TMath::Sqrt(12), 1 / TMath::Sqrt(12));
 
         FairTestDetectorHit* hit = new ((*fHitArray)[ipnt]) FairTestDetectorHit(-1, -1, pos, dpos);
-        hit->AddLink(FairLink(-1, FairRootManager::Instance()->GetEntryNr(), FairRootManager::Instance()->GetBranchId("FairTestDetectorDigi"), ipnt));
+        if ( FairRootManager::Instance() != NULL ) {
```

Comment:
The singleton implementation does not really follow the known rules (private ctor and static instance that create the object by the first call). I started a branch few months ago to correct this (This is also valid for the FairRunXXX classes) but never continue! may be it is time to correct it.
With thread local storage the thread safety issue is solved (FairGenericRootManager.cxx)